### ISSUE TITLE
Error "%s resolves to localhost in a containerized deployment"

### ISSUE THREAD

**Issue by posliedov** [minio/minio#16754]
#### docker-compose.yaml

```yaml
version: '3.7'

services:
  minio:
    container_name: minio
    #image: minio/minio:RELEASE.2023-01-02T09-40-09Z
    image: minio/minio:RELEASE.2023-01-06T18-11-18Z
    network_mode: "host"
    volumes:
      - /data/minio:/data
    environment:
      MINIO_ROOT_USER: minio
      MINIO_ROOT_PASSWORD: pa…rd
      MINIO_SERVER_URL: "https://minio.example.lan"
      MINIO_BROWSER_REDIRECT_URL: "https://minio-admin.example.lan"
    command: server --console-address ":9001" http://minio{01...04}.example.lan/data
    restart: unless-stopped
```

## Expected Behavior
All nodes should start without errors and discover each other

## Current Behavior
Every node has the same error, only the fqdn of the node in the logs is different:
```
Recreating minio ... done
Attaching to minio
minio    |
minio    | API: SYSTEM()
minio    | Time: 19:30:20 UTC 03/03/2023
minio    | Error: minio01.example.lan resolves to localhost in a containerized deployment, waiting for it to resolve to a valid IP (*errors.errorString)
minio    |        host="minio01.example.lan", elapsedTime="1 second elapsed"
minio    |        8: internal/logger/logger.go:258:logger.LogIf()
minio    |        7: internal/logger/logonce.go:104:logger.(*logOnceType).logOnceIf()
minio    |        6: internal/logger/logonce.go:135:logger.LogOnceIf()
minio    |        5: cmd/endpoint.go:489:cmd.Endpoints.UpdateIsLocal()
minio    |        4: cmd/endpoint.go:674:cmd.CreateEndpoints()
minio    |        3: cmd/endpoint-ellipses.go:383:cmd.createServerEndpoints()
minio    |        2: cmd/server-main.go:220:cmd.serverHandleCmdArgs()
minio    |        1: cmd/server-main.go:524:cmd.serverMain()
```

## Possible Solution
Use release `2023-01-02T09-40-09Z` that does not have this problem
(The problem is present from the `2023-01-06T18-11-18Z` version to the current one)

## Your Environment

* Version used (`minio --version`): `2023-02-27T18-10-45Z`
* Server setup and configuration: `Docker 20.10.5`
* Operating System and version (`uname -a`): `Linux minio01.example.lan 5.10.0-10-amd64 #1 SMP Debian 5.10.84-1 (2021-12-08) x86_64 GNU/Linux`
* `cat /etc/hosts` (according to https://www.debian.org/doc/manuals/debian-reference/ch05.en.html#_the_hostname_resolution):
```bash
127.0.0.1	localhost
127.0.1.1	minio01.example.lan	minio01

# The following lines are desirable for IPv6 capable hosts
::1	localhost	ip6-localhost ip6-loopback
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters
```

**Comment by harshavardhana** [comment]
You are doing the setup wrong @posledov we need to know the external IP here to communicate with other servers 

Binding to 127.0.0.1 we have no way to know you want to talk outside the network.

**Comment by posliedov** [comment]
@harshavardhana 
First of all thank you for your reply!

> You are doing the setup wrong

Yes, it's possible, but this configuration **worked fine until the 2023-01-06T18-11-18** release (GH-16356)

> Binding to 127.0.0.1

As far as I understand, this is because:
- `network_mode: "host"` is defined for the container
- the content of `/etc/hosts` is copied inside the container
- `hostname` resolves to 127.0.1.1

## **But**…

In search of a workaround, I did the following experiment:

### Step 1

`/etc/hosts` on the host system:
```sh
127.0.0.1	localhost
127.0.1.1	minio1.example.lan	minio1

# !!! 192.168.254.254 is INCORRECT IP !!!
192.168.254.254	minio1.example.lan	minio1

# The following lines are desirable for IPv6 capable hosts
::1	localhost	ip6-localhost ip6-loopback
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters
```
`192.168.254.254` is incorrect (see below… `ens192` – the only one physical interface):

```console
root@minio1.example.lan /opt/minio # ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
2: ens192: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 00:50:56:bd:06:08 brd ff:ff:ff:ff:ff:ff
    altname enp11s0
    inet 10.13.1.1/24 brd 10.13.1.255 scope global ens192
       valid_lft forever preferred_lft forever
3: docker0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default
    link/ether 02:42:84:e1:b4:04 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.1/16 brd 172.17.255.255 scope global docker0
       valid_lft forever preferred_lft forever

root@minio1.example.lan /opt/minio #
```

### Step 2
Recreate minio container and start it:
```console
root@minio1.example.lan /opt/minio # docker-compose down ; docker-compose up -d ;
Stopping minio ... done
Removing minio ... done
Creating minio ... done

root@minio1.example.lan /opt/minio # docker exec -it minio cat /etc/hosts
127.0.0.1	localhost
127.0.1.1	minio1.example.lan	minio1

# !!! 192.168.254.254 is INCORRECT IP !!!
192.168.254.254	minio1.example.lan	minio1

# The following lines are desirable for IPv6 capable hosts
::1	localhost	ip6-localhost ip6-loopback
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters

root@minio1.example.lan /opt/minio # docker-compose logs minio
Attaching to minio
minio    | Waiting for all MinIO sub-systems to be initialized.. lock acquired
minio    | Automatically configured API requests per node based on available memory on the system: 52
minio    | All MinIO sub-systems initialized successfully in 101.210316ms
minio    | MinIO Object Storage Server
minio    | Copyright: 2015-2023 MinIO, Inc.
minio    | License: GNU AGPLv3 <https://www.gnu.org/licenses/agpl-3.0.html>
minio    | Version: RELEASE.2023-02-27T18-10-45Z (go1.19.6 linux/amd64)
minio    |
minio    | Status:         4 Online, 0 Offline.
minio    | API: https://minio.example.lan
minio    | Console: https://minio-admin.example.lan
minio    |
minio    | Documentation: https://min.io/docs/minio/linux/index.html

root@minio1.example.lan /opt/minio #
```

As you can see, minio started normally, despite the fact that inside the container the `hostname` does not resolve to an existing IP address:

```console
root@minio1.example.lan /opt/minio # docker exec -it minio ifconfig -a
docker0: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        inet 172.17.0.1  netmask 255.255.0.0  broadcast 172.17.255.255
        ether 02:42:84:e1:b4:04  txqueuelen 0  (Ethernet)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

ens192: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 10.13.1.1  netmask 255.255.255.0  broadcast 10.13.1.255
        ether 00:50:56:bd:06:08  txqueuelen 1000  (Ethernet)
        RX packets 7781377  bytes 1701550304 (1.5 GiB)
        RX errors 0  dropped 1391  overruns 0  frame 0
        TX packets 6942224  bytes 786949555 (750.4 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 1314  bytes 1987290 (1.8 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 1314  bytes 1987290 (1.8 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

root@minio1.example.lan /opt/minio # 
```

> we need to know the external IP here to communicate with other servers
> Binding to 127.0.0.1 we have no way to know you want to talk outside the network.

And one more **very important thing**:
```console
root@minio1.example.lan /opt/minio # docker exec -it minio netstat -nltp
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 127.0.0.1:33347         0.0.0.0:*               LISTEN      -
tcp        0      0 127.0.0.1:9000          0.0.0.0:*               LISTEN      1/minio
tcp        0      0 0.0.0.0:80              0.0.0.0:*               LISTEN      -
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN      -
tcp        0      0 0.0.0.0:443             0.0.0.0:*               LISTEN      -
tcp        0      0 0.0.0.0:10050           0.0.0.0:*               LISTEN      -
tcp6       0      0 :::9000                 :::*                    LISTEN      1/minio
tcp6       0      0 :::9001                 :::*                    LISTEN      1/minio
tcp6       0      0 :::22                   :::*                    LISTEN      -
tcp6       0      0 :::10050                :::*                    LISTEN      -

root@minio1.example.lan /opt/minio #
```

As you cen see above, minio binds not only to `127.0.0.1`, but also to `:::9000` (i.e. **to ALL IP addresses** that exist inside the container)!

### Сonclusion of the experiment

So, @harshavardhana , maybe the added check "`%s resolves to localhost in a containerized deployment`" prevents the normal start of minio at a time when minio would be able to start and function normally without it as before (before this check was added)?
